### PR TITLE
Use code description to pass error code document url in lsp

### PR DESCRIPTION
**Pull request by 928PJY**

### CODE REVIEW

File: src/docfx/serve/lsp/LanguageServerHost.cs [hunk 1]
```diff
@@ -39,6 +40,10 @@ internal class LanguageServerHost
                     .AddSingleton<DiagnosticPublisher>()
                     .AddSingleton<LanguageServerCredentialProvider>()
                     .AddSingleton<LanguageServerBuilder>()
+                    .AddSingleton(new ConfigurationItem
```
Comment (yufeih) [inline src/docfx/serve/lsp/LanguageServerHost.cs:43]: What's this for?

Comment (928PJY) [reply src/docfx/serve/lsp/LanguageServerHost.cs:43]: Get the extension config like trace level

**Review by yufeih — APPROVED** [review]

File: src/docfx/lib/log/ErrorList.cs [hunk 2]
```diff
@@ -10,15 +10,23 @@ namespace Microsoft.Docs.Build
 {
     internal class ErrorList : ErrorBuilder, IReadOnlyList<Error>
     {
+        private readonly object _lock = new();
+
         private readonly List<Error> _items = new();
 
-        public Error this[int index] => _items[index];
+        public Error this[int index] => _items.ToArray()[index];
```
Comment (yufeih) [inline src/docfx/lib/log/ErrorList.cs:17]: Must guard every access to `_items` with lock. You also don't need a dedicated `_lock` object, can simply lock `_items`.

File: src/docfx/serve/lsp/LanguageServerHost.cs [hunk 3]
```diff
@@ -18,41 +18,49 @@ internal class LanguageServerHost
...
 
             var builder = server.GetRequiredService<LanguageServerBuilder>();
 
-            await Task.WhenAll(
-                server.WaitForExit,
-                Task.Run(() => builder.Run(cts.Token)));
+            await server.WasStarted;
+            notificationListener?.OnInitialized();
+
+            await Task.Run(() => builder.Run(cancellationToken), cancellationToken);
```
Comment (yufeih) [inline src/docfx/serve/lsp/LanguageServerHost.cs:63]:
```suggestion
            await builder.Run(cancellationToken);
```

File: src/docfx/serve/lsp/LanguageServerHost.cs [hunk 4]
```diff
@@ -18,41 +18,49 @@ internal class LanguageServerHost
...
+                            .AddLogging());
+                    options.OnUnhandledException = (e) =>
+                    {
+                        notificationListener?.OnException(e);
+                        Telemetry.TrackException(e);
+                    };
+                },
+                cancellationToken);
 
             var builder = server.GetRequiredService<LanguageServerBuilder>();
```
Comment (yufeih) [inline src/docfx/serve/lsp/LanguageServerHost.cs:58]: consider move this line to line 62

File: test/docfx.Test/DocfxTest.cs [hunk 5]
```diff
@@ -75,7 +75,8 @@ public static void Run(TestData test, DocfxTestSpec spec)
                 throw new TestSkippedException("OS not supported");
             }
 
-            lock (s_locks.GetOrAdd($"{test.FilePath}-{test.Ordinal:D2}", _ => new object()))
+            var lockKey = test.FilePath.Contains("lsp.yml") ? test.FilePath : $"{test.FilePath}-{test.Ordinal:D2}";
```
Comment (yufeih) [inline test/docfx.Test/DocfxTest.cs:78]: revert?

File: src/docfx/lib/log/ErrorList.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/lib/log/ErrorList.cs:17]:
```csharp
class ErrorList : ErrorBuilder
{
    public void Error[] ToArray();
}
```

File: test/docfx.Test/lsp/LanguageServerTestClient.cs [hunk 6]
```diff
@@ -28,7 +28,10 @@ internal class LanguageServerTestClient : ILanguageServerNotificationListener, I
         private static readonly JsonDiff s_languageServerJsonDiff = CreateLanguageServerJsonDiff();
 
         private readonly string _workingDirectory;
+        private readonly CancellationTokenSource _clientCts = new(30000);
```
Comment (yufeih) [inline test/docfx.Test/lsp/LanguageServerTestClient.cs:31]: try not share this token

File: src/docfx/lib/log/ErrorList.cs [hunk 7]
```diff
@@ -18,12 +17,22 @@ internal class ErrorList : ErrorBuilder, IReadOnlyList<Error>
 
         public override bool HasError => _items.Any(item => item.Level == ErrorLevel.Error);
 
-        public override void Add(Error error) => _items.Add(error);
+        public override void Add(Error error)
```
Comment (yufeih) [inline src/docfx/lib/log/ErrorList.cs:20]: - Remove the `Count` and `this[int index]` method as read operations also needs `lock`.
- lock `HasError` method

File: test/docfx.Test/lsp/LanguageServerTestClient.cs [hunk 8]
```diff
@@ -40,7 +42,9 @@ internal class LanguageServerTestClient : ILanguageServerNotificationListener, I
         private int _clientNotificationSent;
         private int _clientNotificationReceived;
         private int _clientNotificationReceivedBeforeSync; // For expectNoNotification
-        private TaskCompletionSource _notificationSync = new TaskCompletionSource();
+        private TaskCompletionSource _notificationSync = new();
+
+        private CancellationToken DefaultCancellationToken => new CancellationTokenSource(30000).Token;
```
Comment (yufeih) [inline test/docfx.Test/lsp/LanguageServerTestClient.cs:47]: nit: consider turn this into a method. A property usually implies field semantic, which gives the impression that the cancellation token is shared. Consider a more descriptive name like `TimeOut`

**Review by yufeih — APPROVED** [review]
👍

**Pull request merged**